feat(state): Add `BlobPartMessage` and `HasBlobPartMessage` to consensus state and reactor

alesforz opened the pull request:
Part of the work of https://github.com/berachain/cometbft/pull/26.

### Context
Peers exchange `BlobPartMessage` and `HasBlobPartMessage` while gossiping blocks. `BlobPartMessage` contains a fragment of a blob needed for reconstructing it, while `HasBlobPartMessage` allows a peer (A) to inform another peer (B) that it has already received a specific blob part. This prevents redundant transmissions, reducing unnecessary bandwidth usage.

### Changes
This PR introduces the `BlobPartMessage` and `HasBlobPartMessage` types to the consensus state and reactor. However, it does not include logic for handling the blob parts themselves; that will be implemented in a separate PR.

Commit:
Add `BlobPartMessage` in consensus reactor

Commit:
Add `BlobPartMessage` case in consensus state `handleMsg()`

Commit:
add blob part in protobuf messages

Commit:
Add `HasProposalBlobPartMessage` to consensus reactor

Commit:
Implement `Wrap()` method for types `BlobPart` and `HasProposalBlobPart`

Commit:
Added `BlobPartMessage` and `HasBlobPartMessage` to functions
serializing/deserializing protobuf messages to/from their corresponding
Go types

Commit:
Add `ProposalBlobPart` to events

Commit:
Add `BlobParts` to `peerStateStats`.

Does it make sense to mark a peer as good if it has sent enough blob parts?
See `switch` in `Reactor.peerStatsRoutine()`.

alesforz commented:
Does it make sense to add 
```go
cmtjson.RegisterType(&BlobPartMessage{}, "tendermint/BlobPart")
```
[here](https://github.com/alesforz/cometbft/blob/7f3c50019ce1509ec4ce5791bfd05f53be3635bf/internal/consensus/reactor.go#L1756)?
Would this require generating protobufs for the `proto/tendermint` folder?

alesforz commented on internal/consensus/reactor.go:
```diff
@@ -1052,7 +1086,10 @@ func (conR *Reactor) peerStatsRoutine() {
 					conR.Switch.MarkPeerAsGood(peer)
 				}
+			case *BlobPartMessage:
+				// TODO
```
Does it make sense to mark a peer as good if it has sent enough blob parts? Can we omit this?

melekes replied:
We can omit this.

alesforz replied:
Ok, should we still add a call to `RecordBlobPart()`  for stats recording? 
```go
case *BlobPartMessage:
    ps.RecordBlobPart()
```
At a first glance, I can't tell if these stats are used only to mark peers good/bad or they are picked up by the metrics system.

melekes replied:
I think `ps.RecordBlobPart()` is supposed to mark the blob part as received in the peer state. Then it's used to check if we have this blob part or not later if we receive it again.

melekes replied:
Having a metric for the number of blob parts received is a good idea.

Commit:
Add `BlobPartMessage` to WAL replay after crash.

Commit:
Fix unit test `TestMarshalJSONPeerState`

alesforz commented:
The linter is failing, but that's not a real problem.
It's complaining about not using the blob object, but in the next PR(s) the code will use it.

alesforz commented:
Does it make sense to add a `NewValidBlobMessage` [here](https://github.com/alesforz/cometbft/blob/7f3c50019ce1509ec4ce5791bfd05f53be3635bf/internal/consensus/reactor.go#L269)?

alesforz commented:
@melekes @jmalicevic @greg-szabo 
Let me know if you think I've missed parts of the code to which I should add these types 🙂

alesforz commented:
Remade this a draft to check the existing tests

Commit:
Merge branch 'alesforz/feature/blobs' into alesforz/tmp_merge_feature_blobs_into_blob-part-msg

Commit:
Update unit test `TestMsgToProto`

Commit:
Add test `TestBlobPartMessageValidateBasic`

Commit:
Small refactor of `State.readReplayMessage()`

Commit:
Add a blob ID to some WAL tests

melekes commented on internal/consensus/msgs_test.go:
```diff
@@ -21,194 +21,245 @@ import (
+	protoBlkPart, err := blkPart.ToProto()
+	require.NoError(t, err)
+
+	protoBlobpart, err := blobPart.ToProto()
```
```suggestion
	protoBlobPart, err := blobPart.ToProto()
```

melekes reviewed (approved):
LGTM

melekes commented:
> Does it make sense to add a NewValidBlobMessage [here](https://github.com/alesforz/cometbft/blob/7f3c50019ce1509ec4ce5791bfd05f53be3635bf/internal/consensus/reactor.go#L269)?

Can we rely on `NewValidBlockMessage` to avoid introducing new types? I.e., if there's a new valid block, then there **MUST BE** a new valid blob message. I think we can tie them together.

melekes commented:
> Does it make sense to add `cmtjson.RegisterType(&BlobPartMessage{}, "tendermint/BlobPart")`

I think so, yes. What we're using it for? RPC? I mean `cmtjson`

alesforz commented:
> > Does it make sense to add `cmtjson.RegisterType(&BlobPartMessage{}, "tendermint/BlobPart")`
> 
> I think so, yes. What we're using it for? RPC? I mean `cmtjson`

yes, `cmtjson` is our custom pkg handling serialization/deserialization to/from amino json. Wherever we use json in our code, we are using this pkg. RPC as well.

I'm just confused by the `tendermint/BlobPart` string though. 
I'm wondering it this would require generating protobufs for the `proto/tendermint` 🤔

alesforz commented:
> Can we rely on `NewValidBlockMessage` to avoid introducing new types? I.e., if there's a new valid block, then there **MUST BE** a new valid blob message. I think we can tie them together.

What do you have in mind?

Blocks and blobs do not have a strict 1:1 correspondence; a block may exist without an associated blob. Therefore, we don't need a new valid blob message for every new valid block message.

The `NewValidBlockMessage` is defined in the protobuf defs as:
```go
type NewValidBlock struct {
	Height             int64            
	Round              int32            
	BlockPartSetHeader v1.PartSetHeader 
	BlockParts         *v11.BitArray    
	IsCommit           bool             
}
```
So we could add `BlobPartSetHeader` and `BlobParts` in this message. But given that we gossip blobs separately from blocks, we might have a complete block before having the complete blob, so `BlobPartSetHeader` and `BlobParts` might be incomplete 🤔 

BTW, the code sends a `NewValidBlockMessage` in the [commit phase](https://github.com/alesforz/cometbft/blob/b5f9563299e9dd60e351a2f82f1386381f9d72f4/internal/consensus/state.go#L1909) and when there is a [new vote](https://github.com/alesforz/cometbft/blob/b5f9563299e9dd60e351a2f82f1386381f9d72f4/internal/consensus/state.go#L2585) that brings the total votes count to 2/3+.

Commit:
Address comments from code review

melekes commented:
> I'm just confused by the tendermint/BlobPart string though.

It's just an old convention. You can change it to `cometbft/BlobPart`. I don't think it touches protos.

melekes commented:
> BTW, the code sends a NewValidBlockMessage in the [commit phase](https://github.com/alesforz/cometbft/blob/b5f9563299e9dd60e351a2f82f1386381f9d72f4/internal/consensus/state.go#L1909) and when there is a [new vote](https://github.com/alesforz/cometbft/blob/b5f9563299e9dd60e351a2f82f1386381f9d72f4/internal/consensus/state.go#L2585) that brings the total votes count to 2/3+.

Thanks. What would the purpose of `NewValidBlobMessage` then be?

Commit:
Register types `BlobPartMessage` and `HasProposalBlobPartMessage` in consensus
reactor `init`()

alesforz commented:
> Thanks. What would the purpose of `NewValidBlobMessage` then be?

That was my question 😄 I was looking at the code and saw the `NewValidBlockMessage` that peers broadcast once they "see" a valid block, and wondered whether we needed the same for blobs. I think we don't, but wanted to hear more opinions about it 😄

jmalicevic commented on internal/consensus/replay.go:
```diff
@@ -46,48 +49,95 @@ func (cs *State) readReplayMessage(msg *TimedWALMessage, newStepSub types.Subscr
+				"round", msg.Round,
+				"peer", peerID,
+			)
+		} // end switch msg := m.Msg.(type)
```
Is this comment left here by accident?

jmalicevic commented on internal/consensus/reactor.go:
```diff
@@ -453,6 +458,20 @@ func (conR *Reactor) subscribeToBroadcastEvents() {
 		conR.Logger.Error("Error adding listener for events (ProposalBlockPart)", "err", err)
 	}
+
+	err := conR.conS.evsw.AddListenerForEvent(
```
Do you plan to add the FireEvent call for this event in the subsequent PR? Or does it make sense to do it in this PR? 
See https://github.com/cometbft/cometbft/blob/789cf37ba3692a9b8a666c753f66b281dcd46517/internal/consensus/state.go#L2461 and https://github.com/cometbft/cometbft/blob/789cf37ba3692a9b8a666c753f66b281dcd46517/internal/consensus/state.go#L1800 ( I imagine you need only the second case for Blobs).

alesforz replied:
IMHO it should go in the PR that deals with consensus updates. In this PR there is none of that, so there isn't a "right place" in which to place the call yet

jmalicevic replied:
Ok sounds good, can you please in this PR just add a TODO here that there is no corresponding FireEvent. If someone else picks it up, they might overlook this

Commit:
Add `TODO` in consensus reactor `subscribeToBroadcastEvents()`.

In this PR we don't call `FireEvent` for `BlobPartMessage`.
We must add this call in a future PR implementing consensus state updates for
blobs. Because this code isn't there yet, there isn't a "correct spot" to place
the call to `FireEvent` yet.
